src/lightning/fabric/utilities/optimizer.py

```diff
@@ -31,4 +52,104 @@ def _optimizers_to_device(optimizers: Iterable[Optimizer], device: _DEVICE) -> N
 def _optimizer_to_device(optimizer: Optimizer, device: _DEVICE) -> None:
     """Moves the state of a single optimizer to the device."""
     for p, v in optimizer.state.items():
-        optimizer.state[p] = apply_to_collection(v, Tensor, move_data_to_device, device, allow_frozen=True)
+        optimizer.state[p] = apply_to_collection(v, torch.Tensor, move_data_to_device, device, allow_frozen=True)
+
+
+def _no_op() -> None:
+    pass
+
+
+@contextlib.contextmanager
+def _apply_optimizers_during_fsdp_backward(
+    optimizers: Iterable[Optimizer],
+    module_wrapper: "_FabricModule",
+) -> Generator[None, None, None]:
+    """Call `Optimizer.step` as gradients become available.
+
+    NOTE: This is an EXPERIMENTAL utility and exploits behavior which is not
+          part of the FSDP public API. Use at your own risk.
+
+    By moving optimizer step invocation into the backward call we can free
+    gradients earlier and reduce peak memory.
+    """
+    assert SUPPORTS_OPTIMIZER_IN_FSDP_BACKWARD
+    apply_lock = threading.Lock()
+
+    from lightning.fabric.wrappers import _FabricModule
+
+    assert isinstance(module_wrapper, _FabricModule)
+    param_handles = _get_fsdp_handles(module := module_wrapper._forward_module)
+    assert param_handles, f"Module {module} does not appear to contain any FSDP modules."
+    fsdp_state = _get_module_fsdp_state(module)
+    assert fsdp_state is not None
+    fsdp_stream = fsdp_state._streams["post_backward"]
+
+    if isinstance(optimizers, Optimizer):
+        optimizers = [optimizers]
+
+    # We cannot trigger the optimizer step until all parameters are ready.
+    remaining = {}
+    for optimizer in optimizers:
+        unfinished: Dict[torch.nn.Parameter, None] = {}  # Use Dict as an ordered set.
+        for group in optimizer.param_groups:
+            for p in group["params"]:
+                if p not in unfinished:
+                    assert p not in remaining, f"{p=} is shared between two optimizers."
+                    unfinished[p] = None
+                    remaining[p] = (optimizer, unfinished)
+
+    def maybe_step(parameters: Iterable[torch.nn.Parameter], post_step: Callable[[], None] = lambda: None) -> None:
+        for p in tuple(parameters):
+            optimizer, unfinished = remaining.pop(p)
+            unfinished.pop(p)
+            if not unfinished:
+                optimizer.step()
+                optimizer.zero_grad(set_to_none=True)
+                post_step()
```

Comment:
What is a good example of using this hook? Can you add assertions that it gets called as expected?

What if we remove it and the default step implementation is

```python
def step(optimizer):
    optimizer.step()
    optimizer.zero_grad()
```

and the user can do

```python
with _apply_optimizers_during_fsdp_backward(optimizers, model, my_step):
```

Reply:
It's not user facing. It's an implementation detail to free memory when the step runs.